### PR TITLE
Add SpatialAdaptiveAveragePooling.

### PR DESCRIPTION
**Pull request by ruotianluo**
Most of the code is based on SpatialAdaptiveMaxPooling, only changing max pooling to average pooling.

### CODE REVIEW

File: lib/THNN/generic/SpatialAdaptiveAveragePooling.c [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+#pragma omp parallel for private(k)
+  for (k = 0; k < nslices; k++)
+  {
+    /* loop over output */
+    long i, j;
+    for(i = 0; i < oheight; i++)
+    {
+      int y_start = (int)floor((float)i / oheight * iheight);
+      int y_end   = (int)ceil((float)(i + 1) / oheight * iheight);
+      int kH = y_end-y_start;
```
Comment (ruotianluo) [inline lib/THNN/generic/SpatialAdaptiveAveragePooling.c]: Is it better to use
int y_start = (int)floor(float(i * iheight) / oheight);
int y_end   = (int)ceil((float)((i + 1) * iheight) / oheight);
or 
int y_start = i * iheight / oheight;
int y_end = i * iheight / oheight + (i * iheight % oheight)>0?1:0;

The first one is faster and more accurate than the original implementation (but still involves float number calculation). The second is uglier but accurate.

**Comment by fmassa** [comment]
This looked good. Any special reason why you closed it?
Also, do you plan to add a cuda version?

**Comment by ruotianluo** [comment]
I made some little change.  The cuda version is almost ready, I will make the pull requests tonight.

**Comment by ruotianluo** [comment]
#1105 torch/cunn#418

**Pull request closed without merging**